### PR TITLE
[GH-598] chore: update linter config and fix issues

### PR DESCRIPTION
**Pull request by sergiught**
## Description

This PR aims at updating our linter config so we can have better checks for our codebase and ensure high standards. 

Some issues were previously ignored by the linter due to the `exclude-use-default: false` config missing on the `issues` block. Most of the small linter issues were also fixed within this PR, however others are ignored and need to be fixed separately. 

## References

- https://github.com/openfga/openfga/issues/598

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: .golangci.yaml [hunk 1]
```diff
@@ -51,10 +55,23 @@ linters-settings:
...
-    # Do not allow non-required aliases.
-    no-extra-aliases: false
+    no-unaliased: true # Do not allow unaliased imports of aliased packages.
+    no-extra-aliases: false # Do not allow non-required aliases.
     alias:
       - pkg: github.com/openfga/api/proto/openfga/v1
         alias: openfgav1
+
+issues:
+  exclude-use-default: false
```
Comment (sergiught) [inline .golangci.yaml:65]: By not having this configuration set within the linter config, we were ignoring certain issues. For example the `errcheck` linter issues.

File: .golangci.yaml [hunk 2]
```diff
@@ -28,21 +28,25 @@ linters:
...
-    # - perfsprint
-    # - prealloc
-    # - tparallel
+    - revive
+    - godot
+    - bodyclose
+    - gocritic
+    - godot
+    - gocyclo
+    #- gosec Too many different issues. Needs to be tackled in a separate PR.
```
Comment (sergiught) [inline .golangci.yaml:37]: This is important to have as a linter within the config as it gives us insights into potential security issues, however due to the big amount of issues reported this needs to be tackled as a separate and dedicated PR.

File: .golangci.yaml [hunk 3]
```diff
@@ -51,10 +55,23 @@ linters-settings:
...
+      linters:
+        - bodyclose
+  exclude:
+    - "should have a package comment" # stylecheck | revive
+    - "exported (.+) should have comment" # revive
+    - "comment on exported (.+) should be of the form" # revive
+    - "(.+) name will be used as (.+) by other packages, and that stutters; consider calling this (.+)" # revive
+    - "Error return value of (.+) is not checked" # errcheck
+    - "unused-parameter" # revive
+    - "exported func (.+) returns unexported type (.+), which can be annoying to use" # revive
```
Comment (sergiught) [inline .golangci.yaml:77]: Ignoring these for now, as they need to be tackled in dedicated PRs due to the high volume of issues.

File: tests/tests.go [hunk 4]
```diff
@@ -8,17 +8,15 @@ import (
...
 	"github.com/openfga/openfga/cmd/run"
 	serverconfig "github.com/openfga/openfga/internal/server/config"
 	"github.com/openfga/openfga/pkg/logger"
+	"github.com/openfga/openfga/pkg/testfixtures/storage"
+	"github.com/openfga/openfga/pkg/testutils"
 )
 
 // TestClientBootstrapper defines a client interface definition that can be used by tests
-// to bootstrap OpenFGA resources (stores, models, relationship tuples, etc..) needed to
+// to bootstrap OpenFGA resources (stores, models, relationship tuples, etc..) Needed to
```
Comment (miparnisari) [inline tests/tests.go:19]: caps not needed here

File: .golangci.yaml (hunk 2 above)
Comment (miparnisari) [reply .golangci.yaml:37]: what about `contextcheck`, `exhaustive`, `errorlint`, `gochecknoglobals`, `noctx`, `inamedparam`, `persprint`, `prealloc` and `tparallel`? can we keep them as comments? or did you see no value in them?

File: .golangci.yaml [hunk 5]
```diff
@@ -28,21 +28,25 @@ linters:
...
+    - gocyclo
+    #- gosec Too many different issues. Needs to be tackled in a separate PR.
+
 linters-settings:
+  gofmt:
+    simplify: true
+  staticcheck:
+    checks: [ "all" ]
+  gocyclo:
+    min-complexity: 58 # This is our baseline at the moment, but our goal is to take this down to 15.
```
Comment (miparnisari) [inline .golangci.yaml:45]: why 15? seems very arbitrary. Also, are we actually going to work on it? Otherwise this goal is wishful thinking 🤷‍♀️

Comment (sergiught) [reply .golangci.yaml:45]: The concept of cyclomatic complexity was invented by Thomas J. McCabe in 1976 and based on his research:

## Ideal Cyclomatic Complexity any language
### Range Considerations:

- A **cyclomatic complexity between 1 and 10** is typically seen as good and indicates relatively simple code that should be easy to test and maintain.
- A score **above 10** might start to indicate more complex code where it might be beneficial to consider refactoring to simplify methods/functions, especially when it goes above 20. Code with complexity between 10 and 20 is still manageable, but caution should be used as complexity grows.
- A cyclomatic complexity **over 20** is often regarded as a sign of software that could be difficult to understand and maintain. Such complexity levels should prompt a review and potential refactoring, possibly breaking down complex functions into smaller, more manageable pieces.

--- 

@miparnisari 15 seems like a good average based on those ranges. I can remove the comment as most likely we won't work on it right now, but I'd keep track of the complexity so it doesn't go higher than what it currently is, otherwise the codebase will get extremely unmanageable.

Comment (willvedd) [reply .golangci.yaml:45]: I think measuring and enforcing the cyclomatic complexity is well intentioned idea but it is imperfect measure of complexity and even then not a good measure of code _quality_. I'd much prefer our codebase to be readable and high-quality rather than adhering to some linter's rules. This would be a case of the tail wagging the dog.

Comment (miparnisari) [reply .golangci.yaml:45]: i'm okay with tracking the value. Our next step could be reducing the number to say 40 and see where we break this, this would be a good indication of which bits are complex

Comment (sergiught) [reply .golangci.yaml]: @willvedd 
Cyclomatic complexity is **closely tied** with code quality but it should not be used in isolation. It's a metric used as a proxy to assess aspects of code quality, particularly maintainability and testability. 

### Testability
- **Higher Complexity, Lower Testability:** A higher cyclomatic complexity indicates more independent paths through the program, which typically means more test cases are needed to achieve adequate coverage. This can make comprehensive testing more challenging and time-consuming, potentially lowering the overall testability of the code.
- **Test Coverage Metrics:** Cyclomatic complexity directly influences the minimum number of test cases required for thorough testing. This is because each unique path ideally needs to be tested to ensure all logic branches are executed, which helps in detecting bugs or unexpected behavior in those branches.

### Maintainability

- **Complexity and Maintenance Efforts:** More complex code (i.e., code with a higher cyclomatic complexity) can be harder to understand and modify. Each decision point can introduce a potential error or require changes when modifications are made, increasing the maintenance burden.
- **Refactoring Triggers:** High complexity values often signal that a piece of code may need to be refactored into simpler, more manageable components. Refactoring can help reduce cyclomatic complexity by breaking down complex functions into simpler ones, making the code easier to maintain.

### Readability 

- **Indirect Impact:** While cyclomatic complexity does not measure readability directly, highly complex code can be more difficult to read and understand. Code that is hard to understand can be error-prone and may deter modifications by other developers due to the fear of introducing bugs.


--- 

Let's please consider the fact that right now inside this config the value is 58, which is incredibly high and lenient. It means that it's allowing functions that have over 1000 lines of code to exist. Ideally we take this down over time as @miparnisari suggests, e.g. to 40. This is mostly so we can start tracking another useful metric.

File: .golangci.yaml (hunk 2 above)
Comment (sergiught) [reply .golangci.yaml:37]: I honestly felt that they hinder more than aid at this point in time, as they flagged too many issues that aren't really problems. For example we make quite heavy use of global vars through out the codebase and having `gochecknoglobals` would not bring much of a benefit at this point in time. I think it would be kind of frustrating to get bogged down by minor things when you’re in the middle of solving bigger problems. Once we bring the codebase in a better shape we could consider them again, I'd rather not leave them commented out but if you feel strongly about this I can add comments for them, let me know.

**Review by miparnisari — APPROVED** [review]

**Comment by jon-whit** [comment]
@sergiught as a maintainer of the project, I'd like to know how does this change the code I write in this project because it impacts my development on the project. If the majority of @openfga/backend is in favor of merging this, then let's proceed, but let's please document in the description what this means for existing maintainers, because I know it'll come as a surprise to me when certain lint rules are impacting my development that I had not previously seen while working on this project. Could we enumerate what the new linter rules mean please?

---

We as a maintainer team recently discussed to strive to be more inline with the Go standards and not deviate too far unless we had a strong reason to do so, and that seemed to be generally well received and agreed upon. I don't see a strong reason/rationale besides anecdotal and subjective "code quality" presented in this PR to be more opinionated than the Go standards here. What is the purpose of being more opinionated than the standard Go format rules here? Have we been a victim of a problem that our linter could have caught, or are we just being more strict for strictness sake?

As a general note to the @openfga/backend team, I think we should lean on the side of less is more unless we have a strong reason to do so. I think the right approach for our maintainer team would be to start with a discussion of what rules and patterns we'd like to encourage in our codebase and _why_, agree on those, and then introduce the lint rules to enforce those agreed upon rules. I don't think enabling swaths of linter rules all at once without prior art or discussions is a good idea because it changes prior expectations with existing maintainers. It's like a breaking code change for maintainer's expectations, and it creates churn instead of productivity.

File: .golangci.yaml (hunk 5 above)
Comment (miparnisari) [reply .golangci.yaml]: Tangentially related: https://goreportcard.com/report/github.com/openfga/openfga

Why is this not complaining about `gocyclo`? 😆 
image

**Comment by jpadilla** [comment]
The fact that most of the changes required after introducing these rules and config changes are minimal, tells me team/codebase is already pretty well-aligned and that we shouldn't see much impact if any (unless running this is noticeably slower). We could then argue that we don't need them, but enforcement of these specific rules for future contributions doesn't seem like a bad idea 👍

**Review by jpadilla — APPROVED** [review]
